src/predictor/cpu_predictor.cc

```diff
@@ -156,60 +135,78 @@ class CPUPredictor : public Predictor {
+    // tree_begin is not used, right now we just enforce it to be 0.
+    CHECK_EQ(tree_begin, 0);
+    auto* out_preds = &predts->predictions;
+    CHECK_GE(predts->version, tree_begin);
+    if (predts->version == 0) {
+      CHECK_EQ(out_preds->Size(), 0);
+      this->InitOutPredictions(dmat->Info(), out_preds, model);
     }
-    this->InitOutPredictions(dmat->Info(), out_preds, model);
 
-    ntree_limit *= model.learner_model_param_->num_output_group;
-    if (ntree_limit == 0 || ntree_limit > model.trees.size()) {
-      ntree_limit = static_cast<unsigned>(model.trees.size());
+    uint32_t const output_groups =  model.learner_model_param_->num_output_group;
+    CHECK_NE(output_groups, 0);
+    // Right now we just assume ntree_limit provided by users means number of tree layers
+    // in the context of multi-output model
+    uint32_t real_ntree_limit = ntree_limit * output_groups;
+    if (real_ntree_limit == 0 || real_ntree_limit > model.trees.size()) {
+      real_ntree_limit = static_cast<uint32_t>(model.trees.size());
     }
 
-    this->PredLoopInternal(dmat, &out_preds->HostVector(), model,
-                           tree_begin, ntree_limit);
-
-    auto cache_entry = this->FindCache(dmat);
-    if (cache_entry == cache_->cend()) {
-      return;
+    uint32_t const end_version = (tree_begin + real_ntree_limit) / output_groups;
+    // When users have provided ntree_limit, end_version can be lesser, cache is violated
+    if (predts->version > end_version) {
+      CHECK_NE(ntree_limit, 0);
+      this->InitOutPredictions(dmat->Info(), out_preds, model);
+      predts->version = 0;
     }
-    if (cache_entry->second.predictions.Size() == 0) {
-      // See comment in GPUPredictor::PredictBatch.
-      InitOutPredictions(cache_entry->second.data->Info(),
-                         &(cache_entry->second.predictions), model);
-      cache_entry->second.predictions.Copy(*out_preds);
+    uint32_t const beg_version = predts->version;
+    CHECK_LE(beg_version, end_version);
+
+    if (beg_version < end_version) {
+      this->PredInternal(dmat, &out_preds->HostVector(), model,
+                         beg_version * output_groups,
+                         end_version * output_groups);
     }
+
+    // delta means {size of forest} * {number of newly accumulated layers}
+    uint32_t delta = end_version - beg_version;
+    CHECK_LE(delta, model.trees.size());
+    predts->Update(delta);
+
+    CHECK(out_preds->Size() == output_groups * dmat->Info().num_row_ ||
+          out_preds->Size() == dmat->Info().num_row_);
   }
 
   void UpdatePredictionCache(
       const gbm::GBTreeModel& model,
       std::vector<std::unique_ptr<TreeUpdater>>* updaters,
-      int num_new_trees) override {
+      int num_new_trees,
+      DMatrix* m,
+      PredictionCacheEntry* predts) override {
     int old_ntree = model.trees.size() - num_new_trees;
     // update cache entry
-    for (auto& kv : (*cache_)) {
```

Comment:
It might be worth noting, before this function update the cache for all DMatrix, including validation datasets.  Now it only update the training DMatrix, as I don't think the `updater` can/should have information about validation datasets in the future (otherwise that would be peeking and violates the purpose of validation).